lib/adapters/file/write.js

```diff
@@ -0,0 +1,85 @@
+var _ = require('underscore');
+var Juttle = require('../../runtime/index').Juttle;
+var fs = require('fs');
+
+var Write = Juttle.proc.sink.extend({
+    procName: 'write-file',
+    initialize: function(options, params) {
+        var allowed_options = ['file', 'limit'];
+        var unknown = _.difference(_.keys(options), allowed_options);
+        if (unknown.length > 0) {
+            throw this.compile_error('RT-UNKNOWN-OPTION-ERROR', {
+                proc: 'write file',
+                option: unknown[0]
+            });
+        }
+
+        if (!_.has(options, 'file')) {
+            throw this.compile_error('RT-MISSING-OPTION-ERROR', {
+                proc: 'write file',
+                option: "file"
+            });
+        }
+
+        this.limit = options.limit ? options.limit : 100000;
+        this.filename = options.file;
+        this.queue = [];
+    },
+
+    start: function() {
+        this.flush();
+    },
+
+    process: function(points) {
+        this.logger.debug('process', points);
+        this.queue = this.queue.concat(points);
+    },
+
+    flush: function() {
+        if (this.queue.length !== 0) {
+            var data = [];
+            var points = [];
+
+            try {
+                data = fs.readFileSync(this.filename, 'utf8');
+                points = JSON.parse(data);
+            } catch (err) {
+                if (err.code !== 'ENOENT') {
+                    this.logger.error('error reading file:', err.toString());
+                }
+            }
+
+            var space_left = this.limit - points.length;
+            if (space_left < this.queue.length) {
+                points = points.concat(this.queue.slice(0, space_left));
+                // error since we really can't take any more points after this and
+                // should stop the whole pipeline
+                this.trigger('error', this.runtime_error('RT-PROC-LIMIT-EXCEEDED', {
+                    proc: 'write file'
+                }));
+            } else {
+                // lets write out some more points
+                points = points.concat(this.queue);
+            }
+            data = JSON.stringify(points, null, 4);
+            fs.writeFileSync(this.filename, data);
+            this.queue = [];
+        }
+
+        if (this.isdone) {
+            this.done();
+        } else {
+            var self = this;
+            this.program.scheduler.schedule(new Date().getTime() + 1000, function() {
+                self.flush();
+            });
+        }
+    },
+
+    eof: function() {
+        this.logger.debug('eof');
+        this.isdone = true;
```

Comment:
It seems better to me to have `eof` trigger an immediate flush instead of just setting `isdone` and waiting.

This should be trivial to handle now since all the fs operations are async, but as @bkutil pointed out, they should be changed to be async, so you'll need to make sure you don't race between the scheduler callback and the immediate eof from flush. I think it's easy to handle by just setting `isdone` immediately after calling flush so that you know a subsequent callback won't race.